### DISCUSSION TITLE
How to integrate core Marked.js test cases into extension packages?

### DISCUSSION THREAD

**Discussion by calculuschild** [markedjs/marked#3012] [Q&A]
I don't know enough about NPM package development....

Given that Marked.js extensions are a thing now, I know how to write unit tests for the extension itself. However, is it possible to also run all of Marked.js's test cases to ensure an extension isn't breaking default functionality?

I'm looking at https://github.com/calculuschild/marked-extended-tables which completely replaces parsing of GFM tables. In addition to the new functions it adds, it should also still pass all of the table-related tests we already cover in Marked. As Marked gets updates and adds more tests to handle edge cases, the extension falls behind, and can't necessarily use the same fixes, which becomes very tedious to manually track.

Is there some best practice way I can automatically include the latest Marked test suite as part of testing the extension each time I update Marked as a dependency? I have Marked as a dev dependency, so it seems like I should be able to reuse that somehow, but testing has always been my weakness.

**Comment by UziTech** [comment]
You can get the current tests by installing the marked github repo with something like:

```
npm install --save-dev marked-repo@https://github.com/markedjs/marked/tarball/v9.0.3
```

then you will have to create a test framework that uses the files in `.\node_modules\marked-repo\test\specs`

**Reply by UziTech** [reply]
It would be possible to create a script that exports the test framework and allows the user to pass in the `Marked` instance to run the tests against.

**Reply by calculuschild** [reply]
Would there be a way to just re-use the existing Marked test framework to reduce the amount of bloat to the extension package? It's not clear to me what you have in mind for a script that exports a test framework.

**Reply by UziTech** [reply]
I think it should work to just use jasmine to run https://github.com/markedjs/marked/blob/master/test/specs/run-spec.js. You will have to create your own asyncMatchers like https://github.com/markedjs/marked/blob/master/test/helpers/helpers.js. Though creating your own run-spec.js might be better so you can choose which tests to skip.

**Reply by UziTech** [reply]
Ya I think the best way is to create your own https://github.com/markedjs/marked/blob/master/test/specs/run-spec.js. You can still use https://github.com/markedjs/marked/blob/master/test/helpers/load.js to get the actual tests. Then you can use jasmine or jest or any other test framework and run just the tests that you want.

**Reply by UziTech** [reply]
Here is an example of what it could look like to run marked spec tests in an extension: https://github.com/markedjs/marked-bidi/pull/362

**Reply by calculuschild** [reply]
Thanks @UziTech . Your example looks promising.

I'm curious if you would be in support of a future PR that changes `matchers.js` to allow passing in a Marked instance?

**Reply by UziTech** [reply]
Ya my example uses jest but marked uses jasmine which is why I couldn't just use the marked matchers

**Reply by UziTech** [reply]
We could move the tests out of marked into some sort of @markedjs/testutils package so it can just be installed as a devdep and ran with passing in marked.

**Reply by calculuschild** [reply]
Yeah I wouldn't mind something like that.
